(v2) appium install driver deletes previous installed driver(s) when installing multiple drivers.

## The problem

I installed appium 2.0.0-beta.27 via npm, 
first installed uiautomator2 driver - run test with uiautomator2 -> all ok
then installed espresso driver - run test with uiautomator2 again -> error:
```
Could not start a new session. Response code 500. Message: An unknown server-side error occurred while processing the command. Original error: Could not find a driver for automationName 'UIAutomator2' and platformName android'. Have you installed a driver that supports those capabilities? Run 'appium driver list --installed' to see. (Lower-level error: Cannot find module '/home/cmaganu/.appium/node_modules/appium-uiautomator2-driver
```

## Environment

* Appium version (or git revision) that exhibits the issue: 2.0.0-beta.27
* Last Appium version that did not exhibit the issue (if applicable): -
* Desktop OS/version used to run Appium: Ubunntu 20.10
* Node.js version (unless using Appium.app|exe): v16.14.2
* Npm or Yarn package manager: npm 8.5.5
* Mobile platform/version under test: Android 10
* Real device or emulator/simulator: emulator
* Appium CLI or Appium.app|exe: cli

## Details
- after `appium driver install uiautomator2`:
```
cmaganu@cma-x1:~/git/lt-appium-mobile-tests(main)$ ls -Al ~/.appium/node_modules/
total 128
drwxrwxr-x 5 cmaganu cmaganu   4096 mar 28 16:45 appium-uiautomator2-driver
drwxrwxr-x 3 cmaganu cmaganu   4096 mar 28 16:45 .cache
-rw-rw-r-- 1 cmaganu cmaganu 121877 mar 28 16:45 .package-lock.json
```
```
cmaganu@cma-x1:~/git/lt-appium-mobile-tests(main)$ cat  ~/.appium/node_modules/.cache/appium/extensions.yaml 
drivers:
  uiautomator2:
    automationName: UiAutomator2
    platformNames:
      - Android
    mainClass: AndroidUiautomator2Driver
    pkgName: appium-uiautomator2-driver
    version: 2.0.5
    installType: npm
    installSpec: uiautomator2
plugins: {}
schemaRev: 2
```
```
cmaganu@cma-x1:~/git/lt-appium-mobile-tests(main)$ appium driver list --installed
✔ Listing installed drivers
- uiautomator2@2.0.5 [installed (NPM)]
```
- after `appium driver install espresso`:
```
cmaganu@cma-x1:~/git/lt-appium-mobile-tests(main)$ appium driver list --installed
✔ Listing installed drivers
- uiautomator2@2.0.5 [installed (NPM)]
- espresso@2.2.0 [installed (NPM)]
```
```
cmaganu@cma-x1:~/git/lt-appium-mobile-tests(main)$ ls -Al ~/.appium/node_modules/
total 708
drwxrwxr-x 6 cmaganu cmaganu   4096 mar 28 16:49 appium-espresso-driver
drwxrwxr-x 3 cmaganu cmaganu   4096 mar 28 16:45 .cache
-rw-rw-r-- 1 cmaganu cmaganu 714864 mar 28 16:49 .package-lock.json
```
```
cmaganu@cma-x1:~/git/lt-appium-mobile-tests(main)$ cat  ~/.appium/node_modules/.cache/appium/extensions.yaml 
drivers:
  uiautomator2:
    automationName: UiAutomator2
    platformNames:
      - Android
    mainClass: AndroidUiautomator2Driver
    pkgName: appium-uiautomator2-driver
    version: 2.0.5
    installType: npm
    installSpec: uiautomator2
  espresso:
    automationName: Espresso
    platformNames:
      - Android
    mainClass: EspressoDriver
    pkgName: appium-espresso-driver
    version: 2.2.0
    installType: npm
    installSpec: espresso
plugins: {}
schemaRev: 2
```
`.appium/package.json` is empty

--- a/package.json
+++ b/package.json
@@ -35,7 +35,6 @@
     "clean:distfiles": "npx rimraf \"packages/*/build\"",
     "clean:monorepo": "npx rimraf node_modules package-lock.json",
     "clean:packages": "lerna clean -y && npx rimraf \"packages/*/package-lock.json\"",
-    "clean:distfiles": "npx rimraf \"packages/*/build\"",
     "clean:types": "tsc -b --clean",
     "dev": "lerna run --parallel --prefix --concurrency=8 dev",
     "dev:types": "tsc -b --watch",

--- a/package.json
+++ b/package.json
@@ -138,6 +138,7 @@
     "@types/shell-quote": "1.7.1",
     "@types/sinon": "10.0.11",
     "@types/sinon-chai": "3.2.8",
+    "@types/teen_process": "1.16.0",
     "@types/through2": "2.0.36",
     "@types/uuid": "8.3.4",
     "@wdio/types": "7.19.1",

--- a/packages/appium/lib/cli/extension-command.js
+++ b/packages/appium/lib/cli/extension-command.js
@@ -1,11 +1,13 @@
 /* eslint-disable no-console */
 
 import _ from 'lodash';
-import { npm, fs, util } from '@appium/support';
+import path from 'path';
+import { npm, fs, util, env } from '@appium/support';
 import { log, spinWith, RingBuffer } from './utils';
 import { SubProcess } from 'teen_process';
 import { INSTALL_TYPE_NPM, INSTALL_TYPE_GIT, INSTALL_TYPE_GITHUB,
          INSTALL_TYPE_LOCAL } from '../extension/extension-config';
+import { packageDidChange } from '../extension/package-changed';
 
 const UPDATE_ALL = 'installed';
 
@@ -184,14 +186,7 @@ class ExtensionCommand {
       throw new Error(`When using --source=${installType}, must also use --package`);
     }
 
-    if (installType === INSTALL_TYPE_LOCAL) {
-      const msg = `Linking ${this.type} from local path into ${this.config.appiumHome}`;
-      const pkgJsonData = await spinWith(this.isJsonOutput, msg, async () => (
-        await npm.linkPackage(this.config.appiumHome, installSpec))
-      );
-      extData = this.getExtensionFields(pkgJsonData);
-      log(this.isJsonOutput, `Successfully linked ${extData.pkgName} into ${this.config.appiumHome}`);
-    } else if (installType === INSTALL_TYPE_GITHUB) {
+    if (installType === INSTALL_TYPE_GITHUB) {
       if (installSpec.split('/').length !== 2) {
         throw new Error(`Github ${this.type} spec ${installSpec} appeared to be invalid; ` +
                         'it should be of the form <org>/<repo>');
@@ -203,40 +198,44 @@ class ExtensionCommand {
       installSpec = installSpec.replace(/\.git$/, '');
       extData = await this.installViaNpm({ext: installSpec, pkgName: /** @type {string} */(packageName)});
     } else {
-      // at this point we have either an npm package or an appium verified extension
-      // name. both of which will be installed via npm.
-      // extensions installed via npm can include versions or tags after the '@'
-      // sign, so check for that. We also need to be careful that package names themselves can
-      // contain the '@' symbol, as in `npm install @appium/fake-driver@1.2.0`
-      let name, pkgVer;
-      const splits = installSpec.split('@');
-      if (installSpec[0] === '@') {
-        // this is the case where we have an npm org included in the package name
-        [name, pkgVer] = [`@${splits[1]}`, splits[2]];
+      let pkgName, pkgVer;
+      if (installType === INSTALL_TYPE_LOCAL) {
+        pkgName = path.isAbsolute(installSpec) ? installSpec : path.resolve(installSpec);
       } else {
-        // this is the case without an npm org
-        [name, pkgVer] = splits;
-      }
-      let pkgName;
-
-      if (installType === INSTALL_TYPE_NPM) {
-        // if we're installing a named package from npm, we don't need to check
-        // against the appium extension list; just use the installSpec as is
-        pkgName = name;
-      } else {
-        // if we're installing a named appium driver (like 'xcuitest') we need to
-        // dereference the actual npm package ('appiupm-xcuitest-driver'), so
-        // check it exists and get the correct package
-        const knownNames = Object.keys(this.knownExtensions);
-        if (!_.includes(knownNames, name)) {
-          const msg = `Could not resolve ${this.type}; are you sure it's in the list ` +
-                      `of supported ${this.type}s? ${JSON.stringify(knownNames)}`;
-          throw new Error(msg);
+        // at this point we have either an npm package or an appium verified extension
+        // name or a local path. both of which will be installed via npm.
+        // extensions installed via npm can include versions or tags after the '@'
+        // sign, so check for that. We also need to be careful that package names themselves can
+        // contain the '@' symbol, as in `npm install @appium/fake-driver@1.2.0`
+        let name;
+        const splits = installSpec.split('@');
+        if (installSpec[0] === '@') {
+          // this is the case where we have an npm org included in the package name
+          [name, pkgVer] = [`@${splits[1]}`, splits[2]];
+        } else {
+          // this is the case without an npm org
+          [name, pkgVer] = splits;
         }
-        pkgName = this.knownExtensions[name];
-        // given that we'll use the install type in the driver json, store it as
-        // 'npm' now
-        installType = INSTALL_TYPE_NPM;
+
+        if (installType === INSTALL_TYPE_NPM) {
+          // if we're installing a named package from npm, we don't need to check
+          // against the appium extension list; just use the installSpec as is
+          pkgName = name;
+        } else {
+          // if we're installing a named appium driver (like 'xcuitest') we need to
+          // dereference the actual npm package ('appiupm-xcuitest-driver'), so
+          // check it exists and get the correct package
+          const knownNames = Object.keys(this.knownExtensions);
+          if (!_.includes(knownNames, name)) {
+            const msg = `Could not resolve ${this.type}; are you sure it's in the list ` +
+                        `of supported ${this.type}s? ${JSON.stringify(knownNames)}`;
+            throw new Error(msg);
+          }
+          pkgName = this.knownExtensions[name];
+          // given that we'll use the install type in the driver json, store it as
+          // 'npm' now
+          installType = INSTALL_TYPE_NPM;
+        }
       }
 
       extData = await this.installViaNpm({ext, pkgName, pkgVer});
@@ -253,6 +252,11 @@ class ExtensionCommand {
 
     const extManifest = {...extData, installType, installSpec};
     await this.config.addExtension(extName, extManifest);
+
+    // update the if we've changed the local `package.json`
+    if (await env.hasAppiumDependency(this.config.appiumHome)) {
+      await packageDidChange(this.config.appiumHome);
+    }
 
     // log info for the user
     log(this.isJsonOutput, this.getPostInstallText({extName, extData}));

--- a/packages/appium/lib/config.js
+++ b/packages/appium/lib/config.js
@@ -86,7 +86,7 @@ async function getGitRev (useGithubApiFallback = false) {
 /**
  * @param {string} commitSha
  * @param {boolean} [useGithubApiFallback]
- * @returns {Promise<number?>}
+ * @returns {Promise<string?>}
  */
 async function getGitTimestamp (commitSha, useGithubApiFallback = false) {
   const gitRoot = await findGitRoot();

--- a/packages/appium/test/e2e/cli.e2e.spec.js
+++ b/packages/appium/test/e2e/cli.e2e.spec.js
@@ -27,6 +27,8 @@ describe('CLI behavior', function () {
    */
   let appiumHome;
 
+  const testDriverPath = path.dirname(resolveFixture('test-driver/package.json'));
+
   describe('when appium is a dependency', function () {
     /** @type {string} */
     let hashPath;
@@ -142,28 +144,24 @@ describe('CLI behavior', function () {
         });
       });
 
-      describe('when a different driver is installed via appium', function () {
-        const testDriverPath = path.dirname(resolveFixture('test-driver/package.json'));
+      describe('when a different driver is installed via "appium driver install"', function () {
         /** @type {string} */
         let oldHash;
-        /** @type {string} */
-        let oldPkg;
 
         before(async function () {
           await runJson([DRIVER_TYPE, LIST]);
           oldHash = await readHash();
-          oldPkg = await fs.readFile(appiumHomePkgPath, 'utf8');
           await installLocalExtension(appiumHome, DRIVER_TYPE, testDriverPath);
         });
 
-        it('should not update package.json', async function () {
-          const newPkg = await fs.readFile(appiumHomePkgPath, 'utf8');
-          newPkg.should.equal(oldPkg);
+        it('should update package.json', async function () {
+          const newPkg = JSON.parse(await fs.readFile(appiumHomePkgPath, 'utf8'));
+          expect(newPkg).to.have.nested.property('devDependencies.test-driver');
         });
 
-        it('should not update the hash', async function () {
-          const newHash = await fs.readFile(path.join(hashPath), 'utf8');
-          newHash.should.equal(oldHash);
+        it('should update the hash', async function () {
+          const newHash = await readHash();
+          newHash.should.not.equal(oldHash);
         });
 
         it('should update the manifest with the new driver', async function () {
@@ -172,6 +170,11 @@ describe('CLI behavior', function () {
           const manifestParsed = YAML.parse(manifest);
           manifestParsed.should.have.nested.property('drivers.test');
           manifestParsed.should.have.nested.property('drivers.fake');
+        });
+
+        it('should actually install both drivers', function () {
+          expect(() => resolveFrom(appiumHome, '@appium/fake-driver')).not.to.throw;
+          expect(() => resolveFrom(appiumHome, 'test-driver')).not.to.throw;
         });
       });
     });
@@ -283,7 +286,6 @@ describe('CLI behavior', function () {
           ret.uiautomator2.installType.should.eql('npm');
           ret.uiautomator2.installSpec.should.eql('uiautomator2');
           const list = await runList(['--installed']);
-          // @ts-expect-error
           delete list.uiautomator2.installed;
           list.should.eql(ret);
         });
@@ -298,10 +300,40 @@ describe('CLI behavior', function () {
           ret.fake.installType.should.eql('npm');
           ret.fake.installSpec.should.eql('@appium/fake-driver');
           const list = await runList(['--installed']);
-          // @ts-expect-error
           delete list.fake.installed;
           list.should.eql(ret);
         });
+
+        it('should install a driver from npm and a local driver', async function () {
+          await clear();
+          await runInstall([
+            '@appium/fake-driver',
+            '--source',
+            'npm',
+          ]);
+          await installLocalExtension(appiumHome, 'driver', testDriverPath);
+          const list = await runList(['--installed']);
+          expect(list.fake).to.exist;
+          expect(list.test).to.exist;
+          expect(() => resolveFrom(appiumHome, '@appium/fake-driver')).not.to.throw;
+          expect(() => resolveFrom(appiumHome, 'test-driver')).not.to.throw;
+        });
+
+        it('should install _two_ drivers from npm', async function () {
+          await clear();
+          await runInstall([
+            '@appium/fake-driver',
+            '--source',
+            'npm',
+          ]);
+          await runInstall(['appium-uiautomator2-driver', '--source', 'npm']);
+          const list = await runList(['--installed']);
+          expect(list.fake).to.exist;
+          expect(list.uiautomator2).to.exist;
+          expect(() => resolveFrom(appiumHome, '@appium/fake-driver')).not.to.throw;
+          expect(() => resolveFrom(appiumHome, 'appium-uiautomator2-driver')).not.to.throw;
+        });
+
         it('should install a driver from npm with a specific version/tag', async function () {
           const currentFakeDriverVersionAsOfRightNow = '3.0.5';
           await clear();
@@ -311,7 +343,6 @@ describe('CLI behavior', function () {
           ret.fake.installType.should.eql('npm');
           ret.fake.installSpec.should.eql(installSpec);
           const list = await runList(['--installed']);
-          // @ts-expect-error
           delete list.fake.installed;
           list.should.eql(ret);
         });
@@ -328,7 +359,6 @@ describe('CLI behavior', function () {
           ret.fake.installType.should.eql('github');
           ret.fake.installSpec.should.eql('appium/appium-fake-driver');
           const list = await runList(['--installed']);
-          // @ts-expect-error
           delete list.fake.installed;
           list.should.eql(ret);
         });
@@ -345,7 +375,6 @@ describe('CLI behavior', function () {
           ret.fake.installType.should.eql('git');
           ret.fake.installSpec.should.eql(FAKE_DRIVER_DIR);
           const list = await runList(['--installed', '--json']);
-          // @ts-expect-error
           delete list.fake.installed;
           list.should.eql(ret);
         });
@@ -364,7 +393,6 @@ describe('CLI behavior', function () {
             'git+https://github.com/appium/appium-fake-driver',
           );
           const list = await runList(['--installed']);
-          // @ts-expect-error
           delete list.fake.installed;
           list.should.eql(ret);
         });
@@ -377,9 +405,13 @@ describe('CLI behavior', function () {
           ret.fake.installType.should.eql('local');
           ret.fake.installSpec.should.eql(FAKE_DRIVER_DIR);
           const list = await runList(['--installed']);
-          // @ts-expect-error
           delete list.fake.installed;
           list.should.eql(ret);
+
+          // it should be a link!  this may be npm-version dependent, but it's worked
+          // this way for quite awhile
+          const stat = await fs.lstat(path.join(appiumHome, 'node_modules', '@appium', 'fake-driver'));
+          expect(stat.isSymbolicLink()).to.be.true;
         });
       });
 
@@ -528,7 +560,7 @@ describe('CLI behavior', function () {
  * @typedef {import('../../lib/extension/manifest').PluginType} PluginType
  * @typedef {import('../../lib/cli/extension-command').ExtensionListData} ExtensionListData
  * @typedef {import('./e2e-helpers').CliArgs} CliArgs
- * @typedef {import('../../types/types').CliExtensionSubcommand} CliExtensionSubcommand
+ * @typedef {import('../../types/cli').CliExtensionSubcommand} CliExtensionSubcommand
  */
 
 /**

--- a/packages/appium/test/e2e/driver.e2e.spec.js
+++ b/packages/appium/test/e2e/driver.e2e.spec.js
@@ -44,7 +44,7 @@ describe('FakeDriver - via HTTP', function () {
   let appiumHome;
   // since we update the FakeDriver.prototype below, make sure we update the FakeDriver which is
   // actually going to be required by Appium
-  /** @type {import('../../lib/extension/manifest').DriverClass} */
+  /** @type {import('../../types/extension').DriverClass} */
   let FakeDriver;
   /** @type {string} */
   let testServerBaseSessionUrl;
@@ -85,7 +85,7 @@ describe('FakeDriver - via HTTP', function () {
   /**
    *
    * @param {number} port
-   * @param {Partial<import('../../types/types').ParsedArgs>} [args]
+   * @param {Partial<import('../../types/cli').ParsedArgs>} [args]
    */
   async function serverStart (port, args = {}) {
     args = {...args, port, address: TEST_HOST};

--- a/packages/appium/test/e2e/e2e-helpers.js
+++ b/packages/appium/test/e2e/e2e-helpers.js
@@ -21,11 +21,11 @@ export const EXECUTABLE = path.join(PACKAGE_ROOT, 'build', 'lib', 'main.js');
  * Runs the `appium` executable with the given args.
  *
  * If the process exits with a nonzero code, the error will be a {@link AppiumRunError}.
- * @template {import('../../types/types').CliExtensionSubcommand} ExtSubcommand
+ * @template {import('../../types/cli').CliExtensionSubcommand} ExtSubcommand
  * @param {string} appiumHome - Path to `APPIUM_HOME`
  * @param {CliExtArgs<ExtSubcommand> | CliArgs} args - Args, including commands
- * @param {import('@appium/support/lib/npm').TeenProcessExecOpts} [opts] - Options for `teen_process`
- * @returns {Promise<TeenProcessExecResult>}
+ * @param {import('teen_process').ExecOptions} [opts] - Options for `teen_process`
+ * @returns {Promise<import('teen_process').ExecResult<string>>}
  */
 async function run (appiumHome, args, opts = {}) {
   const cwd = PACKAGE_ROOT;
@@ -35,9 +35,6 @@ async function run (appiumHome, args, opts = {}) {
   };
   try {
     args = [...process.execArgv, '--', EXECUTABLE, ...args];
-    /**
-     * @type {TeenProcessExecResult}
-     */
     return await exec(process.execPath, args, {
       cwd,
       env,
@@ -78,7 +75,7 @@ export const runAppium = _.curry(_runAppium);
 
 /**
  * See {@link runAppiumRaw}.
- * @type {AppiumOptsRunner<TeenProcessExecResult>}
+ * @type {AppiumOptsRunner<import('teen_process').ExecResult>}
  */
 async function _runAppiumRaw (appiumHome, args, opts) {
   try {
@@ -90,7 +87,7 @@ async function _runAppiumRaw (appiumHome, args, opts) {
 
 /**
  * Runs the `appium` executable with the given args and returns the entire
- * {@link TeenProcessExecResult} object.
+ * `ExecResult` object.
  *
  * **The third parameter is required**.  Pass empty object if you don't need it.
  **/
@@ -121,7 +118,7 @@ async function _runAppiumJson (appiumHome, args) {
  * i.e., add `@type {MyType}` tag.
  */
 export const runAppiumJson = /**
- * @template {import('../../types/types').CliExtensionSubcommand} ExtSubcommand
+ * @template {import('../../types/cli').CliExtensionSubcommand} ExtSubcommand
  * @type {import('lodash').CurriedFunction2<string, CliExtArgs<ExtSubcommand>|CliArgs, Promise<unknown>>}
  */ (_.curry(_runAppiumJson));
 
@@ -181,13 +178,8 @@ export function formatAppiumArgErrorOutput (stderr) {
  */
 
 /**
- * @typedef {import('@appium/support/lib/npm').TeenProcessExecResult} TeenProcessExecResult
- */
-
-/**
- * @typedef {import('@appium/support/lib/npm').TeenProcessExecError} TeenProcessExecError
- * @typedef {import('@appium/support/lib/npm').TeenProcessExecOpts} TeenProcessExecOpts
  * @typedef {import('../../lib/extension/manifest').ExtensionType} ExtensionType
+ * @typedef {import('@appium/support/lib/npm').TeenProcessExecError} TeenProcessExecError
  * @typedef {import('../../lib/cli/extension-command').ExtensionListData} ExtensionListData
  */
 
@@ -207,7 +199,7 @@ export function formatAppiumArgErrorOutput (stderr) {
 
 /**
  * @template ExtSubCommand
- * @typedef {[import('../../types/types').CliSubcommand, ExtSubCommand, ...string[]]} CliExtArgs
+ * @typedef {[import('../../types/cli').CliSubcommand, ExtSubCommand, ...string[]]} CliExtArgs
  */
 
 /**
@@ -216,7 +208,7 @@ export function formatAppiumArgErrorOutput (stderr) {
 
 /**
  * @template [Result=unknown]
- * @template {import('../../types/types').CliExtensionSubcommand} [ExtSubcommand=never]
+ * @template {import('../../types/cli').CliSubcommand} [ExtSubcommand=never]
  * @callback AppiumRunner
  * @param {string} appiumHome
  * @param {CliExtArgs<ExtSubcommand>|CliArgs} args
@@ -226,10 +218,10 @@ export function formatAppiumArgErrorOutput (stderr) {
 
 /**
  * @template [Result=unknown]
- * @template {import('../../types/types').CliExtensionSubcommand} [ExtSubcommand=never]
+ * @template {import('../../types/cli').CliExtensionSubcommand} [ExtSubcommand=never]
  * @callback AppiumOptsRunner
  * @param {string} appiumHome
  * @param {CliExtArgs<ExtSubcommand>|CliArgs} args
- * @param {import('@appium/support/lib/npm').TeenProcessExecOpts} opts
+ * @param {import('teen_process').ExecOptions} opts
  * @returns {Promise<Result>}
  */

--- a/packages/appium/test/e2e/plugin.e2e.spec.js
+++ b/packages/appium/test/e2e/plugin.e2e.spec.js
@@ -29,7 +29,7 @@ const FAKE_PLUGIN_DIR = path.join(PROJECT_ROOT, 'node_modules', '@appium', 'fake
 describe('FakePlugin', function () {
   /** @type {string} */
   let appiumHome;
-  /** @type {Partial<import('../../types/types').ParsedArgs>} */
+  /** @type {Partial<import('../../types/cli').ParsedArgs>} */
   let baseArgs;
   /** @type {string} */
   let testServerBaseUrl;

--- a/packages/appium/test/fixtures/cli/appium-dependency.package.json
+++ b/packages/appium/test/fixtures/cli/appium-dependency.package.json
@@ -1,6 +1,6 @@
 {
   "name": "my-project",
-  "dependencies": {
+  "devDependencies": {
     "appium": "next"
   }
 }

--- a/packages/appium/types/cli.d.ts
+++ b/packages/appium/types/cli.d.ts
@@ -130,5 +130,5 @@ export type ParsedArgs = DriverOpts &
 export type BuildInfo = {
   version: string;
   'git-sha'?: string;
-  built?: number;
+  built?: string;
 };

--- a/packages/support/lib/npm.js
+++ b/packages/support/lib/npm.js
@@ -2,6 +2,7 @@
 
 import path from 'path';
 import semver from 'semver';
+import { hasAppiumDependency } from './env';
 import { exec } from 'teen_process';
 import { fs } from './fs';
 import * as util from './util';
@@ -28,14 +29,6 @@ export const INSTALL_LOCKFILE_RELATIVE_PATH = path.join(
 );
 
 /**
- * Relative path to lockfile used when linking an extension via `appium`
- */
-export const LINK_LOCKFILE_RELATIVE_PATH = path.join(
-  CACHE_DIR_RELATIVE_PATH,
-  '.link.lock',
-);
-
-/**
  * XXX: This should probably be a singleton, but it isn't.  Maybe this module should just export functions?
  */
 export class NPM {
@@ -49,15 +42,6 @@ export class NPM {
   }
 
   /**
-   * Returns path to "link" lockfile
-   * @private
-   * @param {string} cwd
-   */
-  _getLinkLockfilePath (cwd) {
-    return path.join(cwd, LINK_LOCKFILE_RELATIVE_PATH);
-  }
-
-  /**
    * Execute `npm` with given args.
    *
    * If the process exits with a nonzero code, the contents of `STDOUT` and `STDERR` will be in the
@@ -65,9 +49,9 @@ export class NPM {
    * @param {string} cmd
    * @param {string[]} args
    * @param {ExecOpts} opts
-   * @param {TeenProcessExecOpts} [execOpts]
+   * @param {ExecOpts} [execOpts]
    */
-  async exec (cmd, args, opts, execOpts = {}) {
+  async exec (cmd, args, opts, execOpts = /** @type {ExecOpts} */({})) {
     let { cwd, json, lockFile } = opts;
 
     // make sure we perform the current operation in cwd
@@ -85,10 +69,11 @@ export class NPM {
       runner = async () => await acquireLock(_runner);
     }
 
+    /** @type {import('teen_process').ExecResult<string> & {json?: any}} */
     let ret;
     try {
       const {stdout, stderr, code} = await runner();
-      ret = /** @type {TeenProcessExecResult} */({stdout, stderr, code});
+      ret = {stdout, stderr, code};
       // if possible, parse NPM's json output. During NPM install 3rd-party
       // packages can write to stdout, so sometimes the json output can't be
       // guaranteed to be parseable
@@ -189,18 +174,37 @@ export class NPM {
    * @returns {Promise<import('type-fest').PackageJson>}
    */
   async installPackage (cwd, pkgName, {pkgVer} = {}) {
-    // not only this, this directory needs a 'package.json' inside of it, otherwise, if any
-    // directory in the filesystem tree ABOVE cwd happens to have a package.json or a node_modules
-    // dir in it, NPM will install the module up there instead (silly NPM)
-    const dummyPkgJson = path.resolve(cwd, 'package.json');
-    if (!await fs.exists(dummyPkgJson)) {
-      await fs.writeFile(dummyPkgJson, '{}');
+    /** @type {any} */
+    let dummyPkgJson;
+    const dummyPkgPath = path.join(cwd, 'package.json');
+    try {
+      dummyPkgJson = JSON.parse(await fs.readFile(dummyPkgPath, 'utf8'));
+    } catch (err) {
+      if (err.code === 'ENOENT') {
+        dummyPkgJson = {};
+        await fs.writeFile(dummyPkgPath, JSON.stringify(dummyPkgJson, null, 2), 'utf8');
+      } else {
+        throw err;
+      }
     }
 
+    /**
+     * If we've found a `package.json` containined the `appiumCreated` property,
+     * then we can do whatever we please with it, since we created it.  This is
+     * likely when `APPIUM_HOME` is the default (in `~/.appium`).  In that case,
+     * we want `--global-style` to avoid deduping, and we also do not need a
+     * `package-lock.json`.
+     *
+     * If we _haven't_ found such a key, then this `package.json` isn't a
+     * "dummy" and is controlled by the user.  So we'll just add it as a dev
+     * dep; whatever else it does is up to the user's npm config.
+     */
+    const installOpts = await hasAppiumDependency(cwd) ?
+      ['--save-dev'] :
+      ['--save-dev', '--save-exact', '--global-style', '--no-package-lock'];
+
     const res = await this.exec('install', [
-      '--no-save',
-      '--global-style',
-      '--no-package-lock',
+      ...installOpts,
       pkgVer ? `${pkgName}@${pkgVer}` : pkgName
     ], {
       cwd,
@@ -227,46 +231,6 @@ export class NPM {
       throw new Error('The package was not downloaded correctly; its package.json ' +
                       'did not exist or was unreadable. We looked for it at ' +
                       pkgJsonPath);
-    }
-  }
-
-  /**
-   * @todo: I think this can be an `install` instead of a `link`.
-   * @param {string} cwd
-   * @param {string} pkgPath
-   */
-  async linkPackage (cwd, pkgPath) {
-    // from the path alone we don't know the npm package name, so we need to
-    // look in package.json
-    let pkgName;
-    try {
-      pkgName = require(path.resolve(pkgPath, 'package.json')).name;
-    } catch {
-      throw new Error('Could not find package.json inside the package path ' +
-                      `provided: ${pkgPath}`);
-    }
-
-    // this is added to handle commands with relative paths
-    // ie: "node . driver install --source=local ../fake-driver"
-    pkgPath = path.resolve(process.cwd(), pkgPath);
-
-    // call link with --no-package-lock to ensure no corruption while installing local packages
-    const args = [
-      '--global-style',
-      '--no-package-lock',
-      pkgPath
-    ];
-    const res = await this.exec('link', args, {cwd, lockFile: this._getLinkLockfilePath(cwd)});
-    if (res.json && res.json.error) {
-      throw new Error(res.json.error);
-    }
-
-    // now ensure it was linked to the correct place
-    try {
-      return require(resolveFrom(cwd, `${pkgName}/package.json`));
-    } catch {
-      throw new Error('The package was not linked correctly; its package.json ' +
-                      'did not exist or was unreadable');
     }
   }
 
@@ -301,32 +265,11 @@ export const npm = new NPM();
 // THESE TYPES SHOULD BE IN TEEN PROCESS, NOT HERE
 
 /**
- * Result from a non-zero-exit execution of `appium`
- * @typedef TeenProcessExecResult
- * @property {string} stdout - Stdout
- * @property {string} stderr - Stderr
- * @property {number?} code - Exit code
- * @property {any} json - JSON parsed from stdout
- */
-
-/**
  * Extra props `teen_process.exec` adds to its error objects
  * @typedef TeenProcessExecErrorProps
  * @property {string} stdout - STDOUT
  * @property {string} stderr - STDERR
  * @property {number?} code - Exit code
- */
-
-/**
- * Options unique to `teen_process.exec`. I probably missed some
- * @typedef TeenProcessExecExtraOpts
- * @property {number} [maxStdoutBufferSize]
- * @property {number} [maxStderrBufferSize]
- */
-
-/**
- * All options for `teen_process.exec`
- * @typedef {import('child_process').SpawnOptions & TeenProcessExecExtraOpts} TeenProcessExecOpts
  */
 
 /**

--- a/packages/support/lib/util.js
+++ b/packages/support/lib/util.js
@@ -454,9 +454,10 @@ async function toInMemoryBase64 (srcPath, opts = {}) {
  * longer present on the system. This allows for preventing concurrent behavior across processes
  * using a known lockfile path.
  *
+ * @template T
  * @param {string} lockFile The full path to the file used for the lock
  * @param {LockFileOptions} opts
- * @returns {(behavior: () => Promise<void>) => Promise<void>} async function that takes another async function defining the locked
+ * @returns async function that takes another async function defining the locked
  * behavior
  */
 function getLockFileGuard (lockFile, opts = {}) {
@@ -469,6 +470,10 @@ function getLockFileGuard (lockFile, opts = {}) {
   const check = B.promisify(_lockfile.check);
   const unlock = B.promisify(_lockfile.unlock);
 
+  /**
+   * @param {(...args: any[]) => T} behavior
+   * @returns {Promise<T>}
+   */
   const guard = async (behavior) => {
     let triedRecovery = false;
     do {

